ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java

```diff
@@ -54,91 +58,110 @@ public static RemoteInferenceInputDataSet processInput(MLInput mlInput, Connecto
-                    });
-                } catch (PrivilegedActionException e) {
-                    throw new RuntimeException(e);
-                }
-            }
-            inputData = RemoteInferenceInputDataSet.builder().parameters(processedParameters).build();
+            inputData = processTextDocsInput((TextDocsInputDataSet) mlInput.getInputDataset(), connector, parameters, scriptService);
         } else if (mlInput.getInputDataset() instanceof RemoteInferenceInputDataSet) {
             inputData = (RemoteInferenceInputDataSet)mlInput.getInputDataset();
         } else {
             throw new IllegalArgumentException("Wrong input type");
         }
         if (inputData.getParameters() != null) {
             Map<String, String> newParameters = new HashMap<>();
-            inputData.getParameters().entrySet().forEach(entry -> {
-                if (entry.getValue() == null) {
-                    newParameters.put(entry.getKey(), entry.getValue());
-                } else if (StringUtils.isJson(entry.getValue())) {
+            inputData.getParameters().forEach((key, value) -> {
+                if (value == null) {
+                    newParameters.put(key, null);
+                } else if (org.opensearch.ml.common.utils.StringUtils.isJson(value)) {
                     // no need to escape if it's already valid json
-                    newParameters.put(entry.getKey(), entry.getValue());
+                    newParameters.put(key, value);
                 } else {
-                    newParameters.put(entry.getKey(), escapeJson(entry.getValue()));
+                    newParameters.put(key, escapeJson(value));
                 }
             });
             inputData.setParameters(newParameters);
         }
         return inputData;
     }
+    private static RemoteInferenceInputDataSet processTextDocsInput(TextDocsInputDataSet inputDataSet, Connector connector, Map<String, String> parameters, ScriptService scriptService) {
+        List<String> docs = new ArrayList<>(inputDataSet.getDocs());
+        Optional<ConnectorAction> predictAction = connector.findPredictAction();
+        if (predictAction.isEmpty()) {
+            throw new IllegalArgumentException("no predict action found");
+        }
+        String preProcessFunction = predictAction.get().getPreProcessFunction();
+        preProcessFunction = preProcessFunction == null ? MLPreProcessFunction.DEFAULT_EMBEDDING_INPUT : preProcessFunction;
+        if (MLPreProcessFunction.contains(preProcessFunction)) {
+            Map<String, Object> buildInFunctionResult = MLPreProcessFunction.get(preProcessFunction).apply(docs);
+            return RemoteInferenceInputDataSet.builder().parameters(convertScriptStringToJsonString(buildInFunctionResult)).build();
+        } else {
+            if (preProcessFunction.contains("${parameters")) {
+                StringSubstitutor substitutor = new StringSubstitutor(parameters, "${parameters.", "}");
```

Comment:
Looks like we are using this `${parameters.` in multiple places, what do you think about using a Static variable for this?

Reply:
this is existing code just moved to different place, beside this there are many other places using this string, we should make it a static variable but maybe in next PR